Marlin/src/gcode/feature/pause/M600.cpp

```diff
@@ -43,73 +48,154 @@
  *  L[distance] - Extrude distance for insertion (positive value) (manual reload)
  *  B[count]    - Number of times to beep, -1 for indefinite (if equipped with a buzzer)
  *
+ * Spool_Swap : Tool change after filament runout
+ *
+ *  T[int]: 0=disable : 1/2/3 Enable the max number of swapping
+ *
  *  Default values are used for omitted arguments.
  *
  */
 void GcodeSuite::M600() {
-
-  #if ENABLED(HOME_BEFORE_FILAMENT_CHANGE)
+	
+  #if ENABLED(PAUSE_SPOOL_SWAP) 
+    //extern bool swap_spool_enabled;
+	//extern int16_t swap_spool_stop;
+	const int swap_spool_tmp = parser.intval('T',(FIL_RUNOUT_SENSORS));             
+    
+	// M600 T0/T1/2/3/4 Enable/Disable     
+	if (swap_spool_tmp<FIL_RUNOUT_SENSORS){
+	  if (swap_spool_tmp>0){
+	    // Activate or deactivate if next extruder available
+	    swap_spool_enabled = true ;
+		//Apply max number of swapping 
+		swap_spool_stop = swap_spool_tmp;
+	  }
+	  // Deactivate in all case if no next extruders available
+	  else swap_spool_enabled=false; 
+	  // Applying number or next extruders
+	  
+	  return ;  
+    }      
+  #endif
+	
+  #if ENABLED(PAUSE_HOME_BEFORE)
     // Don't allow filament change without homing first
     if (axis_unhomed_error()) home_all_axes();
   #endif
 
   // Initial retract before move to filament change position
   const float retract = parser.seen('E') ? parser.value_axis_units(E_AXIS) : 0
-    #ifdef PAUSE_PARK_RETRACT_LENGTH
-      - (PAUSE_PARK_RETRACT_LENGTH)
+    #ifdef PAUSE_RET
+      - (PAUSE_RET)
     #endif
   ;
 
   // Lift Z axis
   const float z_lift = parser.linearval('Z', 0
-    #ifdef PAUSE_PARK_Z_ADD
-      + PAUSE_PARK_Z_ADD
+    #ifdef PAUSE_Z_ADD
+      + PAUSE_Z_ADD
     #endif
   );
 
   // Move XY axes to filament exchange position
   const float x_pos = parser.linearval('X', 0
-    #ifdef PAUSE_PARK_X_POS
-      + PAUSE_PARK_X_POS
+    #ifdef PAUSE_X_POS
+      + PAUSE_X_POS
+	#else
+      + current_position[X_AXIS]
     #endif
   );
   const float y_pos = parser.linearval('Y', 0
-    #ifdef PAUSE_PARK_Y_POS
-      + PAUSE_PARK_Y_POS
-    #endif
+    #ifdef PAUSE_Y_POS
+      + PAUSE_Y_POS
+	#else
+      + current_position[Y_AXIS]
+    #endif	
   );
-
+  
   // Unload filament
-  const float unload_length = parser.seen('U') ? parser.value_axis_units(E_AXIS) : 0
-    #if defined(FILAMENT_CHANGE_UNLOAD_LENGTH) && FILAMENT_CHANGE_UNLOAD_LENGTH > 0
-      - (FILAMENT_CHANGE_UNLOAD_LENGTH)
-    #endif
-  ;
-
+  #if ENABLED(PAUSE_SPOOL_SWAP)
+      float unload_length = parser.seen('U') ? parser.value_axis_units(E_AXIS) : 0
+      #ifdef PAUSE_UNLOAD
+        - PAUSE_UNLOAD ;
+      if (swap_spool_enabled) unload_length = -SPOOL_SWAP_UNLOAD;
+      #endif 
```

Comment:
Forced to make like this 
I re use , the M600 code , and my feature is an extension 
To add my feature , sometimes i have to make this way 
It's now serious to make an uneeded computation , but we are in PAUSE and computation is not a problem